src/main/java/org/jenkinsci/plugins/workflow/graph/FlowNode.java

```diff
@@ -244,66 +248,123 @@ public String getUrl() throws IOException {
+        } else {
+            return getMaybeTransientAction(type);
+        }
+    }
+
+    private synchronized void loadActions() {
+        if (actions != null) {
+            return; //Mutation while we acquired lock
+        }
+        try {
+            actions = new CopyOnWriteArrayList<Action>(exec.loadActions(this));
+        } catch (IOException e) {
+            LOGGER.log(WARNING, "Failed to load actions for FlowNode id=" + id, e);
+            actions = new CopyOnWriteArrayList<Action>();
+        }
+    }
 
-    So we create a separate transient field and store List of them there, and intercept every mutation.
- */
     @Exported
     @Override
-    public synchronized List<Action> getActions() {
-                if (actions==null) {
-                    try {
-                        actions = new CopyOnWriteArrayList<Action>(exec.loadActions(this));
-                    } catch (IOException e) {
-                        LOGGER.log(WARNING, "Failed to load actions for FlowNode id=" + id, e);
-                        actions = new CopyOnWriteArrayList<Action>();
-                    }
-                }
+    @SuppressFBWarnings(value = "UG_SYNC_SET_UNSYNC_GET", justification = "CopyOnWrite ArrayList, and field load & modification is synchronized")
+    public List<Action> getActions() {
+        if (actions==null) {
+            loadActions();
+        }
+        /*
+        We can't use Actionable#actions to store actions because they aren't transient,
+        and we need to store actions elsewhere because this is the only mutable part of FlowNode.
 
+        So we create a separate transient field and store List of them there, and intercept every mutation.
+        */
         return new AbstractList<Action>() {
-            @Override
-            public Action get(int index) {
-                return actions.get(index);
-            }
 
-            @Override
-            public void add(int index, Action element) {
-                actions.add(index, element);
-                persist();
-            }
+                @Override
+                public Action get(int index) {
+                    return actions.get(index);
+                }
 
-            @Override
-            public Action remove(int index) {
-                Action old = actions.remove(index);
-                persist();
-                return old;
-            }
+                @Override
+                public void add(int index, Action element) {
+                    actions.add(index, element);
+                    persistSafe();
+                }
 
-            @Override
-            public Action set(int index, Action element) {
-                Action old = actions.set(index, element);
-                persist();
-                return old;
-            }
+                @Override
+                public boolean add(Action value) {
+                    actions.add(value);
+                    persistSafe();
```

Comment:
Get in the habit of running

    git diff master

to look for gratuitous hunks that could be reverted after design changes.